golos.publication/golos.publication.cpp

```diff
@@ -560,63 +647,76 @@ void publication::paymssgrwrd(structures::mssgid message_id) {
-            get_memo("author", message_id));
-    payto(
-            message_id.author, 
-            eosio::asset(payout.amount - token_payout, payout.symbol), 
-            static_cast<enum_t>(payment_t::VESTING), 
-            get_memo("author", message_id));
-
-    tables::vote_table vote_table(_self, message_id.author.value);
-    auto votetable_index = vote_table.get_index<"messageid"_n>();
-    auto vote_itr = votetable_index.lower_bound(std::make_tuple(mssg_itr->id, INT64_MAX));
-    for (auto vote_etr = votetable_index.end(); vote_itr != vote_etr;) {
-       auto& vote = *vote_itr;
-       ++vote_itr;
-       if (vote.message_id != mssg_itr->id) break;
-       vote_table.erase(vote);
-    }
+    auto max_add_payouts = mssg_itr->beneficiaries.size() + 2; //beneficiaries, author tokens and author vesting
+    bool last_payment = completed && ((vesting_payouts.size() + max_add_payouts <= config::max_payments_per_trx) || vesting_payouts.empty());
+
+    if (last_payment) {
+        auto actual_curation_payout = paid + mssg_itr->paid_amount;
+        auto unclaimed_rewards = curation_payout - actual_curation_payout;
+        eosio::check(unclaimed_rewards >= 0, "publication::payrewards: unclaimed_rewards < 0");
+        
+        payout.amount -= curation_payout;
+        
+        int64_t ben_payout_sum = 0;
+        for (auto& ben: mssg_itr->beneficiaries) {
+            auto ben_payout = cyber::safe_pct(payout.amount, ben.weight);
+            eosio::check((0 <= ben_payout) && (ben_payout <= payout.amount - ben_payout_sum), 
+                    "LOGIC ERROR! publication::payrewards: wrong ben_payout value");
+            if (ben_payout > 0) {
+                vesting_payouts.push_back({ben.account, eosio::asset(ben_payout, payout.symbol), get_memo("benefeciary", message_id)});
+                ben_payout_sum += ben_payout;
+            }
+        }
+        
+        payout.amount -= ben_payout_sum;
+        elaf_t token_prop(elai_t(mssg_itr->tokenprop) / elai_t(config::_100percent));
+        auto token_payout = int_cast(elai_t(payout.amount) * token_prop);
+        eosio::check(payout.amount >= token_payout, "publication::payrewards: wrong token_payout value");
+        if (token_payout > 0) {
+            pay_to({{message_id.author, eosio::asset(token_payout, payout.symbol), get_memo("author", message_id)}}, false);
+        }
+        if (payout.amount - token_payout > 0) {
+            vesting_payouts.push_back({message_id.author, eosio::asset(payout.amount - token_payout, payout.symbol), get_memo("author", message_id)});
+        }
+        pay_to(std::move(vesting_payouts), true);
+        
+        tables::vote_table vote_table(_self, message_id.author.value);
+        auto votetable_index = vote_table.get_index<"messageid"_n>();
+        auto vote_itr = votetable_index.lower_bound(std::make_tuple(mssg_itr->id, INT64_MAX));
+        
+        for (auto vote_etr = votetable_index.end(); vote_itr != vote_etr;) {
```

Comment:
the votes are already erased in `fill_curator_payouts`, why do we need to erase here?

Reply:
fill_curator_payouts doesn't erase votes with claim == 0
but send_deletevotes_trx should be called here instead of directly erasing, thank you for paying attention to this.